File: perf-hook-sandhook/src/main/java/com/xander/performance/hook/sandhook/XC_MethodSandHook.java
package com.xander.performance.hook.sandhook;

import com.xander.performance.hook.core.MethodHook;

import de.robv.android.xposed.XC_MethodHook;

public class XC_MethodSandHook extends XC_MethodHook {
  private static final String TAG = "XC_MethodSandHook";

  /**
   * 最外层的 hook 方法回调
   */
  MethodHook methodHook;

  /**
   * 实现外层的 hook param ，用来桥接 epic 的 MethodHookParam
   */
  MethodSandHookParam methodSandHookParam;

  public XC_MethodSandHook(MethodHook methodHook) {
    super();
    this.methodHook = methodHook;
  }

  @Override
  protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
    super.beforeHookedMethod(param);
    if (null == methodSandHookParam) {
      methodSandHookParam = new MethodSandHookParam();
    }
    methodSandHookParam.setMethodHookParam(param);
    methodHook.beforeHookedMethod(methodSandHookParam);
  }

  @Override
  protected void afterHookedMethod(MethodHookParam param) throws Throwable {
    super.afterHookedMethod(param);
    if (null == methodSandHookParam) {
      methodSandHookParam = new MethodSandHookParam();
    }
    methodSandHookParam.setMethodHookParam(param);
    methodHook.afterHookedMethod(methodSandHookParam);
  }
}

File: hook-sandhook/src/main/java/io/github/xanderwang/hook/sandhook/XC_MethodSandHook.kt
package io.github.xanderwang.hook.sandhook

import de.robv.android.xposed.XC_MethodHook
import io.github.xanderwang.hook.core.MethodHook

/**
 * @param hookCallback 最外层的 hook 方法回调
 */
class XC_MethodSandHook(var hookCallback: MethodHook?) : XC_MethodHook() {
    /**
     * 实现外层的 hook param ，用来桥接 epic 的 MethodHookParam
     */
    private val methodSandHookParam by lazy { MethodSandHookParam() }

    @Throws(Throwable::class)
    override fun beforeHookedMethod(param: MethodHookParam) {
        super.beforeHookedMethod(param)
        methodSandHookParam.methodHookParam = param
        hookCallback?.beforeHookedMethod(methodSandHookParam)
    }

    @Throws(Throwable::class)
    override fun afterHookedMethod(param: MethodHookParam) {
        methodSandHookParam.methodHookParam = param
        hookCallback?.afterHookedMethod(methodSandHookParam)
        super.afterHookedMethod(param)
    }

    companion object {
        private const val TAG = "XC_MethodSandHook"
    }
}